cpp/src/components/trim.cu

```diff
@@ -0,0 +1,186 @@
+       cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+       std::optional<edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+       bool do_expensive_check)
+{
+
+    rmm::device_uvector<edge_t> core_numbers(graph_view.number_of_vertices(), handle.get_stream());
+
+    core_number(
+      handle, graph_view, core_numbers.data(), k_core_degree_type_t::OUT, size_t{1}, size_t{1});
+
+   auto out_one_core_first =
+      thrust::make_transform_iterator(core_numbers.begin(), is_one_or_greater_t<edge_t>{});
+
+    rmm::device_uvector<uint8_t> out_one_core_flags(core_numbers.size(), handle.get_stream());
+
+    thrust::copy(handle.get_thrust_policy(),
+                 out_one_core_first,
+                 out_one_core_first + core_numbers.size(),
+                 out_one_core_flags.begin());
+
+    edge_src_property_t<decltype(graph_view), uint8_t> edge_src_out_one_cores(handle,
+                                                                                 graph_view);
+    edge_dst_property_t<decltype(graph_view), uint8_t> edge_dst_out_one_cores(handle,
+                                                                                 graph_view);
+    update_edge_src_property(
+      handle, graph_view, out_one_core_flags.begin(), edge_src_out_one_cores);
+    update_edge_dst_property(
+      handle, graph_view, out_one_core_flags.begin(), edge_dst_out_one_cores);
+
+    rmm::device_uvector<size_t> d_subgraph_offsets_out(2, handle.get_stream());
+    std::vector<size_t> h_subgraph_offsets_out{{0, out_one_core_flags.size()}};
+    raft::update_device(d_subgraph_offsets_out.data(),
+                      h_subgraph_offsets_out.data(),
+                      h_subgraph_offsets_out.size(),
+                      handle.get_stream());
+    graph_t<vertex_t, edge_t, false, multi_gpu> subgraph_out(handle);
+    std::optional<edge_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, weight_t>> subgraph_edge_weights_out{};
+    std::optional<rmm::device_uvector<vertex_t>> renumber_map_out{std::nullopt};
+
+    std::tie(subgraph_out, subgraph_edge_weights_out, 
+      std::ignore, std::ignore, renumber_map_out) = 
+    create_graph_from_edgelist<vertex_t,
+                               edge_t,
+                               weight_t,
+                               edge_t,
+                               int32_t,
+                               false,
+                               multi_gpu>(handle, out_one_core_flags, edge_src_out_one_cores, edge_dst_out_one_cores, edge_weight_view);
+
+    auto subgraph_out_view = subgraph_out.view(); 
+    auto [src, dst, wgt, offsets] = extract_induced_subgraphs(
+    handle,
+    subgraph_out_view,
+    subgraph_edge_weights_out,
+    raft::device_span<size_t const>{d_subgraph_offsets_out.data(), d_subgraph_offsets_out.size()},
+    raft::device_span<vertex_t const>{out_one_core_flags.data(), out_one_core_flags.size()},
+    do_expensive_check);
```

Comment:
We're first extracting a one core (based on the out-degree) from the original graph.

Then, you create a new graph (from the edges that belong to the one core).

Then, extract a one core (based on the in-degree) from the new graph.

This code's return value should be based on the original vertex's renumbering, so you unrenumber at this end.

Reply:
Please ping me if this is confusing.